core/drivers/versal_nvm.c

```diff
@@ -0,0 +1,944 @@
+{
+	size_t i = 0;
+
+	cmd->data[i++] = NVM_API_ID(efuse);
+	if (arg)
+		cmd->data[i++] = *arg;
+
+	if (!ibufs[0].buf)
+		return TEE_SUCCESS;
+
+	cmd->data[i++] = virt_to_phys(ibufs[0].buf);
+	cmd->data[i++] = virt_to_phys(ibufs[0].buf) >> 32;
+
+	for (i = 0; i < MAX_IPI_BUF; i++) {
+		cmd->ibuf[i].mem.alloc_len = ibufs[i].len;
+		cmd->ibuf[i].mem.buf = ibufs[i].buf;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result efuse_req(enum versal_nvm_api_id efuse,
+			    struct versal_nvm_buf *ibufs, uint32_t *arg)
+{
+	TEE_Result ret = TEE_SUCCESS;
+	struct ipi_cmd cmd = { };
+
+	ret = prepare_cmd(&cmd, efuse, ibufs, arg);
+	if (ret)
+		return ret;
+
+	ret = versal_mbox_notify(&cmd, NULL, NULL);
+	if (ret)
+		EMSG("Mailbox error");
+
+	return ret;
+}
+
+static TEE_Result versal_nvm_read(struct versal_nvm_read_req *req)
+{
+	uint32_t *arg = NULL;
+	uint32_t val = 0;
+
+	if (!req)
+		return TEE_ERROR_GENERIC;
+
+	switch (req->efuse_id) {
+	case EFUSE_READ_DNA:
+	case EFUSE_READ_DEC_EFUSE_ONLY:
+	case EFUSE_READ_PUF_SEC_CTRL:
+	case EFUSE_READ_BOOT_ENV_CTRL:
+	case EFUSE_READ_SEC_CTRL:
+	case EFUSE_READ_MISC_CTRL:
+	case EFUSE_READ_SEC_MISC1:
+	case EFUSE_READ_USER_FUSES:
+	case EFUSE_READ_PUF_USER_FUSES:
+	case EFUSE_READ_PUF:
+		break;
+	case EFUSE_READ_OFFCHIP_REVOCATION_ID:
+		val = req->offchip_id;
+		arg = &val;
+		break;
+	case EFUSE_READ_REVOCATION_ID:
+		val = req->revocation_id;
+		arg = &val;
+		break;
+	case EFUSE_READ_IV:
+		val = req->iv_type;
+		arg = &val;
+		break;
+	case EFUSE_READ_PPK_HASH:
+		val = req->ppk_type;
+		arg = &val;
+		break;
+	case BBRAM_READ_USER_DATA:
+		break;
+	default:
+		return TEE_ERROR_GENERIC;
+	}
+
+	return efuse_req(req->efuse_id, req->ibuf, arg);
+}
+
+static TEE_Result versal_nvm_write(struct versal_nvm_write_req *req)
+{
+	uint32_t *arg = NULL;
+	uint32_t val = 0;
+
+	switch (req->efuse_id) {
+	case BBRAM_WRITE_AES_KEY:
+		val = req->bbram.aes_key_len;
+		arg = &val;
+		break;
+	case BBRAM_WRITE_USER_DATA:
+		val = req->bbram.user_data;
+		arg = &val;
+		break;
+	case EFUSE_PUF_USER_FUSE_WRITE:
+	case EFUSE_WRITE_PUF:
+	case EFUSE_WRITE:
+		break;
+	default:
+		return TEE_ERROR_GENERIC;
+	}
+
+	return efuse_req(req->efuse_id, req->ibuf, arg);
+}
+
+TEE_Result versal_efuse_read_dna(uint32_t *buf, size_t len)
+{
+	uint8_t lbuf[1024] __aligned_efuse = { 0 };
+	struct versal_nvm_read_req req = {
+		.efuse_id = EFUSE_READ_DNA,
+	};
+
+	if (len < EFUSE_DNA_LEN)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	req.ibuf[0].buf = lbuf;
+	req.ibuf[0].len = sizeof(lbuf);
+
+	if (versal_nvm_read(&req))
+		return TEE_ERROR_GENERIC;
+
+	memcpy(buf, lbuf, EFUSE_DNA_LEN);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result versal_efuse_read_user_data(uint32_t *buf, size_t len,
+				       uint32_t first, size_t num)
+{
+	uint8_t lbuf[1024] __aligned_efuse = { 0 };
+	struct versal_efuse_user_data cfg __aligned_efuse = {
+		.addr = (uintptr_t)lbuf,
+		.start = first,
+		.num = num, /* fuses needs to be at least 40 bytes */
```

Comment:
return with an error if less than 40 bytes?

Reply:
I'll just remove the comment. the firmware handle function might change and this requirement is not in the API...was just sort of soemething I noticed